### DISCUSSION TITLE
[SOLVED] Documentation & Testing Tools

### DISCUSSION THREAD

**Discussion by zicklag** [RAUI-labs/raui#44] [Q&A] [answered]
Just opening this thread to share different documentation and testing tools that might be useful or worth discussing.

**Comment by zicklag** [comment]
Just found this which helps you test markdown files similar to Rustdoc: https://github.com/budziq/rust-skeptic.

Might not be needed after using Cargo readme, but good to keep in mind.

**Comment by zicklag** [comment]
Whoa, I just found another really nifty trick: https://doc.rust-lang.org/rustdoc/documentation-tests.html#include-items-only-when-collecting-doctests.

![image](https://user-images.githubusercontent.com/25393315/118040107-c5055900-b336-11eb-9576-d844682a3601.png)

**Reply by PsichiX** [reply]
just in time while i'm pushing new version xD
ok, we can play with it and push it after release :D

**Comment by zicklag** [comment, marked as answer]
Hey, I just managed to find a way to test all of the Rust snippets in the website in a way that doesn't require _all_ snippets to come from a full-blown cargo project.

Now in the site source we write:

```md
{% rustdoc_test() %}
```rust
# extern crate raui;
# use raui::prelude::*;
let tree = widget! {
    (content_box)
};
{% end %}
```

And then we can run `just website-doc-tests` and it will go and test all of the code snippets in the site! The update will be comming in a PR with the next section of the guide once I finish it.

It turns out it simple to get rustdoc to test any markdown file you want:

```bash
cargo build --features all -p raui
for file in $(find site/content/ -name '*.md'); do \
    echo "Testing: $file"; \
    rustdoc --crate-name docs-test $file --test -L target/debug/deps; \
done
```

**Reply by PsichiX** [reply]
whoah :O

**Comment by PsichiX** [comment]
hmm, i think you've solved all aspects of the problem with CI docs testing :D is there a way to somehow mark this discussion as closed? (by that i mean it won't pop out on top of the discussion board) 🤔 for now i think of adding `[SOLVED]` in the title.

**Reply by zicklag** [reply]
I moved it to the Q&A category that has "Answers" enabled and then marked an answer. :+1:
